File: examples/quickstart.py
# SPDX-License-Identifier: AGPL-3.0-only
"""
Splyntra Quickstart Example
============================
Demonstrates time-to-first-trace in under 5 minutes.

Prerequisites:
    docker compose up -d   # Start Splyntra
    pip install splyntra   # Install SDK

Run:
    python quickstart.py
    
Then open http://localhost:3000/traces to see your trace.
"""

from splyntra import Splyntra, trace_agent, trace_tool, trace_llm

# 1. Initialize Splyntra (one line)
splyntra = Splyntra(
    api_key="splyntra_dev_key",
    project="quickstart-demo",
)


# 2. Instrument your agent
@trace_agent(name="research_agent", workflow="summarize")
def research_agent(query: str) -> str:
    """A simple agent that searches and summarizes."""
    # Step 1: Plan
    plan = plan_research(query)

    # Step 2: Search
    results = search_web(plan["search_query"])

    # Step 3: Summarize
    summary = summarize(results)

    return summary


@trace_llm(model="gpt-4o", provider="openai")
def plan_research(query: str) -> dict:
    """Simulate an LLM planning step."""
    # In production, this calls your LLM
    import time
    time.sleep(0.1)  # Simulate latency
    return {
        "search_query": f"latest research on {query}",
        "usage": {"prompt_tokens": 150, "completion_tokens": 45},
    }


@trace_tool(name="web.search")
def search_web(query: str) -> list:
    """Simulate a tool call."""
    import time
    time.sleep(0.2)  # Simulate API call
    return [
        {"title": "Result 1", "snippet": "Relevant information..."},
        {"title": "Result 2", "snippet": "More details..."},
    ]


@trace_llm(model="gpt-4o-mini", provider="openai")
def summarize(results: list) -> dict:
    """Simulate summarization."""
    import time
    time.sleep(0.15)
    return {
        "content": "Here is a summary of the research findings...",
        "usage": {"prompt_tokens": 320, "completion_tokens": 88},
    }


# 3. Run it
if __name__ == "__main__":
    print("Running research agent...")
    result = research_agent("AI agent observability")
    print(f"Result: {result}")
    print("\n✓ Trace sent! View at http://localhost:3000/traces")

    # Clean shutdown
    splyntra.shutdown()

File: examples/quickstart.ts
// SPDX-License-Identifier: FSL-1.1-ALv2
/**
 * Splyntra TypeScript / JavaScript Quickstart
 * ===========================================
 * Time-to-first-trace in under five minutes from a Node app.
 *
 * Uses function wrappers (wrapAgent / wrapTool / wrapLLM), which work in both
 * TypeScript and plain JavaScript — no decorators or build config required.
 *
 * Prerequisites:
 *     docker compose up -d        # start Splyntra
 *     npm install @splyntra/sdk
 *
 * Run:
 *     npx tsx examples/quickstart.ts        # TypeScript
 *     # plain JS: rename imports to require() and `node quickstart.js`
 *
 * Then open http://localhost:3000/traces to see your trace.
 */

import { Splyntra, wrapAgent, wrapTool, wrapLLM } from "@splyntra/sdk";

// 1. Initialize Splyntra (one line).
const splyntra = new Splyntra({
  apiKey: "splyntra_dev_key",
  project: "quickstart-ts",
  framework: "custom",
});

// 2. Wrap your agent's steps. wrapLLM reads `usage` for token/cost analytics.
const planResearch = wrapLLM(async (query: string) => {
  await new Promise((r) => setTimeout(r, 100)); // simulate latency
  return { searchQuery: `latest research on ${query}`, usage: { prompt_tokens: 150, completion_tokens: 45 } };
}, "gpt-4o", "openai");

const searchWeb = wrapTool(async (q: string) => {
  await new Promise((r) => setTimeout(r, 200));
  return [{ title: "Result 1", snippet: "Relevant info..." }];
}, "web.search");

const summarize = wrapLLM(async (_results: unknown) => {
  await new Promise((r) => setTimeout(r, 150));
  return { content: "Summary of findings.", usage: { prompt_tokens: 320, completion_tokens: 88 } };
}, "gpt-4o-mini", "openai");

const researchAgent = wrapAgent(async (query: string) => {
  const plan = await planResearch(query);
  const results = await searchWeb(plan.searchQuery);
  const summary = await summarize(results);
  return summary.content;
}, "research_agent", "summarize");

// 3. Run it.
async function main() {
  console.log("Running research agent...");
  const result = await researchAgent("AI agent observability");
  console.log(`Result: ${result}`);
  console.log("\n✓ Trace sent! View at http://localhost:3000/traces");
  await splyntra.shutdown(); // flush before the process exits
}

main();
